### ISSUE TITLE
Added missing support for S/MIME envloped signing

### ISSUE THREAD

**Issue by enigma969** [bbottema/simple-java-mail#252]
Hi, 

I have troubles getting the content of an email that is encrypted and signed (see attachment). The email itself only contains one smime.p7m attachment.

I use following code: 

```java
FileInputStream f = new FileInputStream(new File("email.eml"));

FileInputStream key = new FileInputStream(new File("keystore.p12"));
FileInputStream key2 = new FileInputStream(new File("keystore.p12"));

SmimeKeyStore smimeKeyStore = new SmimeKeyStore(key, "pw".toCharArray());
String alias = smimeKeyStore.getPrivateKeyAliases().iterator().next();

Pkcs12Config yourPkcs12Config = Pkcs12Config.builder()
        .pkcs12Store(key2) // path, File or InputStream
        .storePassword("pw")
        .keyAlias(alias)
        .keyPassword("pw")
        .build();

Email mergedEmail = EmailConverter.emlToEmail(f, yourPkcs12Config);
List<AttachmentResource> list = mergedEmail.getDecryptedAttachments();

for (AttachmentResource r : list) {
    play.Logger.info(r.readAllData());
}
```

But the output of `r.readAllData()` gives me binary data only. 

When I open the email in Thunderbird it is decrypted correctly, it contains text only. 

Can you please help me to get the actual content? 
[email.txt](https://github.com/bbottema/simple-java-mail/files/4267048/email.txt)

**Comment by bbottema** [comment]
Hi, my apologies for the late reply, but I'm currently unable to look into this as I have recently become father to our second child. Takes rather lot of attention I'm finding out. On top of that, the topic here is a rather complicated and needs some proper debugging and research.

I'm always open to suggestions and pull requests though. Sorry I don't have better news for the moment...

**Comment by enigma969** [comment]
Congratulations!!! The customer has to wait then lol :)

**Comment by bbottema** [comment]
When I open this file in Thunderbird, I don't see any attachments...

![image](https://user-images.githubusercontent.com/2115718/87256582-43ce0300-c494-11ea-84b6-139b0584ec2d.png)

**Comment by enigma969** [comment]
Probably because you cannot decrypt it? There is definitely an attachment in it.

**Comment by enigma969** [comment]
I'm wrong, sorry. 

If you look at the content of the file (email.txt) you can see a _Content-Disposition: attachment; filename="smime.p7m"_ part. 

That's the actual content (= text) of the email. But it's encrypted. 

Thunderbird is able to the email without problems and shows the content (private key provided). The library is not.  

You can understand the problem now?

**Comment by bbottema** [comment]
Unless you provide the private key along with the encrypted email, I can't do a root-cause analyses without the private key. The sample emails included in Simple Java Mail's tests all work fine.

Can you provide the private key? Perhaps email it to me directly so I can analyse? I won't publish it in the GIT repo.

**Comment by bbottema** [comment]
I have been provided with a matching keystore and have been analysing the issue. It turns out that S/MIME support was incomplete for the backwards compatible legacy signing protocol: "smime-type=signed-data". Nowadays signing is done using certificate added to an attachment, but legacy style is that the entire attachment content is wrapped similarly to how encrypted content still is wrapped (enveloped content).

I've fixed the support (that was pretty complex!), but I've got it working nicely now. Wrapping up on the code quality and then I'll be shipping out a new release...
